### ISSUE TITLE
Wildcard import "..." is not defined

### ISSUE THREAD

**Issue by qumusabel** [microsoft/pyright#1644]
**Describe the bug**
Pyright throws `reportUndefinedVariable` when trying to access names improrted with a wildcard import.

**To Reproduce**
1. Install a module that makes use of wildcard import.
2. `from module import *`
3. Try to access a member of the module

**Expected behavior**
Pyright doesn't throw an error.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from pwn import *
print(
    p64(0x1337)  # returns bytes, but is marked as undefined by Pyright
)
```

**VS Code extension or command-line**
`coc-pyright` plugin for `coc.nvim`

**Additional context**
I am trying to use [pwntools](/Gallopsled/pwntools) which recommends using a wildcard import so that you won't have to access its `pwnlib` backend direcly, which would result in huge paths.

**Comment by erictraut** [comment]
First, I highly recommend against using wildcard imports, especially when importing from a third-party external library. It pollutes the namespace of your module in unpredictable ways and makes for fragile code. It's one of those features in Python that should have probably never been added.

[PEP 8](https://www.python.org/dev/peps/pep-0008/) states:
>Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools.

Looking at `pwn` and `pwntools`, I don't see the symbol `p64` defined in any top-level namespace. It appears as a class variable in the `MemLeak` class (in `memleak.py`), as a method in the `ELF` class (in `elf.py`) and a method in the `tube` class (in `tube.py`). It's possible that it's exported as a symbol from a native library, but pyright wouldn't have any visibility into that without type stubs.

So I think pyright is doing the right thing here. I don't see anything unexpected.

**Comment by qumusabel** [comment]
> Looking at pwn and pwntools, I don't see the symbol p64 defined in any top-level namespace.

`p64` is defined in `pwn.pwnlib.util.packing`, which is a very long path to use, and it becomes even more clunky if you try, for example, `pwn.pwnlib.tubes.process`, `pwn.pwnlib.util.fiddling.xor`, etc. This is why the authors recommend using a wildcard import which would pull all the basic utility into the namespace. 

> First, I highly recommend against using wildcard imports, 

I absolutely do agree that wildcard imports are bad in 99.9% of the cases, but here it is necessary.

Is it possible to silence the error for the entire file?

**Comment by erictraut** [comment]
If the `pwn` module exports the symbol (as it must, assuming it's available via a wildcard import), then you can use the form `from pwn import p64`. No need to use a wildcard.

I don't see `p64` mentioned anywhere in [pwn.pwnlib/util.packing](https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/util/packing.py).

You can silence the error by adding the following comment somewhere in the file:
```python
# pyright: reportUndefinedVariable=false
```

**Comment by qumusabel** [comment]
This is how the wildcard import implemented, I suppose [pwn.toplevel.py](/Gallopsled/pwntools/blob/dev/pwn/toplevel.py)

> I don't see p64 mentioned anywhere in pwn.pwnlib.util.packing.

There is a function that makes a packer for the specific size, and `p64` seems to be generated at import time. ([pwn.pwnlib.util.packing](/Gallopsled/pwntools/blob/dev/pwnlib/util/packing.py#L290))

Is is possible to generate stubs for this kind of import?

**Comment by erictraut** [comment]
I just noticed another problem. The module `pwn/toplevel.py` is using a dynamic expression to define `__all__`. Pyright (and other static type checkers) do not support such dynamic statements. The supported ways that `__all__` can be defined and manipulated are [documented here](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md#library-interface).

If you comment out this line in `pwn/toplevel.py`, pyright (and the Python interpreter) will import all symbols as part of the wildcard import.

It looks like someone else has already commented out this line in the `pwn` repo. See [here](https://github.com/Gallopsled/pwntools/blob/ec9dfe108b85ac47983e9a98808fcdbb50cb0cdd/pwn/toplevel.py#L91). So presumably this will improve in the next release of `pwn`.

I still strongly recommend against using wildcard imports. There's no good justification for their use when importing from a library.

**Comment by erictraut** [comment]
You can certainly write your own type stubs that define these symbols. If you do, I recommend contributing them back to the `pwn` repo so they are bundled with the next release. That way, everyone benefits from them!

**Comment by qumusabel** [comment]
Thank you very much! I think I'm going to try to write the stubs.

**Comment by qumusabel** [comment]
I've written the stubs file, and pyright seems to have picked it up. But when I removed `.pyi` from the package's directory, the information didn't disappear. Now, for some reason, it remembers some classes but still marks functions like `p64` as unefined. Is there a way to find out where pyright takes definitions from and purge that info? I've searched across my home directory, and there aren't any `.pyi` files for the package except the ones I wrote myself.

**Comment by erictraut** [comment]
Pyright internally caches type information. It attempts to install file watchers and flush its caches when files change, but file watchers are unreliable on some platforms. You may need to manually restart pyright's language server after making a change to a package installed in your site-packages. You can do this by quitting and relaunching your editor or using the "Pyright: Restart Service" command from within your client.

If you want to know where a particular symbol is defined, you can place the cursor over the symbol and invoke the "Find Declaration" command in the editor. This should take you directly to the location of the declaration (whether it's in a ".pyi" or ".py" file).

**Comment by qumusabel** [comment]
> You may need to manually restart pyright's language server

I think this is what removed the `p64` definition from the cache, but the classes' definitions are still there.

> invoke the "Find Declaration" command in the editor

I'm sorry, I can't find the command in the docs/sources. Is there an LSP command that would do the same? I'm using `coc-pyright` and `coc.nvim`.

**Comment by qumusabel** [comment]
Setting `python.analysis.useLibraryCodeForTypes` to `false` I get the same old behaviour. Nothing found except some constants explicitly defined in init files... This probably means that it doesn't need the stubs file, which makes me sad and wondering why it didn't work before...

**Comment by erictraut** [comment]
I don't know if/how coc.nvim exposes the "Find Declaration" capability. This capability is available in most editors that support LSPs, including VS Code.

Note that symbols are resolved on a per-module basis. When pyright is resolving a submodule, it looks first for a "<submodule>.pyi" file. If it finds one, it uses the stub file as a source for symbols exported by that submodule. If it doesn't find a "<submodule>.pyi" file, it then either gives up (if "useLibraryCodeForTypes" is false and the package is not marked "py.typed") or it looks for a "<submodule>.py" file and uses it to determine which symbols are exported from the submodule.

I'm assuming that your plan was to introduce a "packing.pyi" file that provides type information for that submodule. Or was your plan to provide stub files for all of the submodules in this package?

**Comment by qumusabel** [comment]
My idea was to write `__init__.pyi` in the top level package, so that everything's known after a wildcard import. This is inefficient (the file is about 25000 lines long w/docstrings), but it worked as a proof of concept. I thought of organizing stubs by the modules, but I'm not sure if they would still be used 3-4 levels into subimports of `import pwn`.

**Comment by erictraut** [comment]
It should work 3-4 levels down because each submodule is resolved and analyzed independently from other submodules. This is how [Partial Stub Packages](https://www.python.org/dev/peps/pep-0561/#partial-stub-packages) (defined in PEP 561) work.
